base/slicearray.jl

```diff
@@ -0,0 +1,224 @@
+ [1, 2, 3]
+ [4, 5, 6]
+ [7, 8, 9]
+```
+"""
+@inline function eachslice(A; dims, drop=true)
+    _eachslice(A, dims, drop)
+end
+
+"""
+    eachrow(A::AbstractVecOrMat)
+
+Create a [`Rows`](@ref) that indexes over the rows of a vector or matrix `A`.
+
+See also [`eachcol`](@ref) and [`eachslice`](@ref).
+
+!!! compat "Julia 1.1"
+     This function requires at least Julia 1.1.
+
+!!! compat "Julia 1.7"
+     Prior to Julia 1.7, this returned an iterator.
+
+# Example
+
+```jldoctest
+julia> a = [1 2; 3 4]
+2×2 Matrix{Int64}:
+ 1  2
+ 3  4
+
+julia> S = eachrow(a)
+2-element Rows{Matrix{Int64}, CartesianIndices{1, Tuple{Base.OneTo{Int64}}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}:
+ [1, 2]
+ [3, 4]
+
+julia> S[1]
+2-element view(::Matrix{Int64}, 1, :) with eltype Int64:
+ 1
+ 2
+```
+"""
+eachrow(A::AbstractMatrix) = _eachslice(A, (1,), true)
+eachrow(A::AbstractVector) = eachrow(reshape(A, size(A,1), 1))
+
+"""
+    eachcol(A::AbstractVecOrMat)
+
+Create a [`Columns`](@ref) that iterates over the second dimension of matrix `A`, returning the
+columns as `AbstractVector` views.
+
+See also [`eachrow`](@ref) and [`eachslice`](@ref).
+
+!!! compat "Julia 1.1"
+     This function requires at least Julia 1.1.
+
+!!! compat "Julia 1.7"
+     Prior to Julia 1.7, this returned an iterator.
+
+# Example
+
+```jldoctest
+julia> a = [1 2; 3 4]
+2×2 Matrix{Int64}:
+ 1  2
+ 3  4
+
+julia> S = eachcol(a)
+2-element Columns{Matrix{Int64}, CartesianIndices{1, Tuple{Base.OneTo{Int64}}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}}:
+ [1, 3]
+ [2, 4]
+
+julia> S[1]
+2-element view(::Matrix{Int64}, :, 1) with eltype Int64:
+ 1
+ 3
+```
+"""
+eachcol(A::AbstractMatrix) = _eachslice(A, (2,), true)
+eachcol(A::AbstractVector) = eachcol(reshape(A, size(A,1), 1))
+
+"""
+    Rows{M,CI,S}
+
+A special case of [`SliceArray`](@ref) that is a vector of row slices of a matrix, as
+constructed by [`eachrow`](@ref).
+
+[`parent(S)`](@ref) can be used to get the underlying matrix.
+"""
+const Rows{P<:AbstractMatrix,CI,S<:AbstractVector} = SliceArray{1,(1,nothing),P,CI,S}
+
+"""
+    Columns{M,CI,S}
+
+A special case of [`SliceArray`](@ref) that is a vector of column slices of a matrix, as
+constructed by [`eachcol`](@ref).
+
+[`parent(S)`](@ref) can be used to get the underlying matrix.
+"""
+const Columns{P<:AbstractMatrix,CI,S<:AbstractVector} = SliceArray{1,(nothing,1),P,CI,S}
+
+
+IteratorSize(::Type{SliceArray{N,L,P,CI,S}}) where {N,L,P,CI,S} = IteratorSize(CI)
+size(s::SliceArray) = size(s.cartiter)
+size(s::SliceArray, dim) = size(s.cartiter, dim)
+length(s::SliceArray) = length(s.cartiter)
+
+@inline function _slice_index(s::SliceArray{N,L,P,CI,S}, I...) where {N,L,P,CI,S}
+    c = s.cartiter[I...]
```

Comment:
Is there any case where this _isn't_ a no-op?  Or, rather, can't we just do:

```suggestion
@inline function _slice_index(s::SliceArray{N,L,P,CI,S}, c::Vararg{Int,N}) where {N,L,P,CI,S}
```

We'd of course still need to store the axes (instead of the cartiter), but shouldn't this be exactly the same performance?  And simpler?

Reply:
Good point. I'm not sure why I did it this way...

Reply:
I totally get it: you want an iterator over the locations; that's the `CartesianIndices`.  Arrays just bring those along as part of their indexing (if they're not IndexLinear, anyhow).  Took me a minute to see, too.